crates/rpc/src/handlers/validator.rs

```diff
@@ -42,6 +44,56 @@ impl ValidatorData {
     }
 }
 
+#[derive(Debug, Serialize)]
+struct ValidatorBalance {
+    #[serde(with = "serde_utils::quoted_u64")]
+    index: u64,
+    #[serde(with = "serde_utils::quoted_u64")]
+    balance: u64,
+}
+
+fn build_validator_balances(
+    validators: &[Validator],
+    filter_ids: Option<&Vec<ValidatorID>>,
+) -> Vec<ValidatorBalance> {
+    let mut allowed_indices: HashSet<u64> = HashSet::new();
+    let mut allowed_addresses: HashSet<PubKey> = HashSet::new();
+
+    if let Some(ids) = filter_ids {
+        for id in ids {
+            match id {
+                ValidatorID::Index(idx) => {
+                    allowed_indices.insert(*idx);
+                }
+                ValidatorID::Address(pk) => {
+                    if validators.iter().any(|v| &v.pubkey == pk) {
+                        allowed_addresses.insert(pk.clone());
+                    }
+                }
+            }
+        }
+    }
+
+    let mut seen: HashSet<u64> = HashSet::new();
+    let mut balances = Vec::new();
+
+    for (i, validator) in validators.iter().enumerate() {
+        let idx = i as u64;
+        let include = filter_ids.is_none()
+            || allowed_indices.contains(&idx)
+            || allowed_addresses.contains(&validator.pubkey);
+
+        if include && seen.insert(idx) {
+            balances.push(ValidatorBalance {
+                index: idx,
+                balance: validator.effective_balance,
+            });
+        }
+    }
+
+    balances
+}
```

Comment:
```suggestion
fn build_validator_balances(
    validators: &[Validator],
    filter_ids: Option<&Vec<ValidatorID>>,
) -> Vec<ValidatorBalance> {
    let Some(filter_ids) = filter_ids else {
        return vec![];
    };

    let indexes = filter_ids
        .into_iter()
        .filter_map(|validator_id| match validator_id {
            ValidatorID::Index(index) => Some(*index),
            ValidatorID::Address(public_key) => validators
                .iter()
                .enumerate()
                .find(|(_, validator)| &validator.pubkey == public_key)
                .map(|(index, _)| index as u64),
        })
        .collect::<HashSet<_>>();

    let mut balances = vec![];
    for index in indexes {
        let Some(validator) = validators.get(index as usize) else {
            continue;
        };
        balances.push(ValidatorBalance {
            index,
            balance: validator.effective_balance,
        });
    }

    balances
}
```

We can simplify this function quite a bit as seen above

Reply:
In this case it isn't actually handling the case when the query.id is None, that is supposed to return all validators if i'm not wrong, which isn't the case with the code proposed.
Also this implementation runs at time complexity of O(n*m) whereas the previous one is O(n+m).
Let me know your thoughts about these pointers

Reply:
> @KolbyML In this case it isn't actually handling the case when the query.id is None, that is supposed to return all validators if i'm not wrong, which isn't the case with the code proposed. Also this implementation runs at time complexity of O(n*m) whereas the previous one is O(n+m). Let me know your thoughts about these pointers

The previous implementation isn't O(n+m), I am not sure where you are getting that from.

```diff
diff --git a/crates/rpc/src/handlers/validator.rs b/crates/rpc/src/handlers/validator.rs
index 3cc8a47..dc990da 100644
--- a/crates/rpc/src/handlers/validator.rs
+++ b/crates/rpc/src/handlers/validator.rs
@@ -1,10 +1,10 @@
+use std::collections::HashSet;
+
 use actix_web::{
     HttpResponse, Responder, get, post,
     web::{Data, Json, Path, Query},
 };
-use std::collections::HashSet;
 use ream_bls::PubKey;
-
 use ream_consensus::validator::Validator;
 use ream_storage::db::ReamDB;
 use serde::{Deserialize, Serialize};
@@ -53,45 +53,25 @@ struct ValidatorBalance {
 }
 
 fn build_validator_balances(
-    validators: &[Validator],
+    validators: &[(Validator, u64)],
     filter_ids: Option<&Vec<ValidatorID>>,
 ) -> Vec<ValidatorBalance> {
-    let mut allowed_indices: HashSet<u64> = HashSet::new();
-    let mut allowed_addresses: HashSet<PubKey> = HashSet::new();
-
-    if let Some(ids) = filter_ids {
-        for id in ids {
-            match id {
-                ValidatorID::Index(idx) => {
-                    allowed_indices.insert(*idx);
-                }
-                ValidatorID::Address(pk) => {
-                    if validators.iter().any(|v| &v.pubkey == pk) {
-                        allowed_addresses.insert(pk.clone());
-                    }
-                }
+    let filtered_ids = filter_ids.map(|filter_ids| filter_ids.iter().collect::<HashSet<_>>());
+    validators
+        .iter()
+        .enumerate()
+        .filter(|(index, (validator, _))| match &filtered_ids {
+            Some(ids) => {
+                ids.contains(&ValidatorID::Index(*index as u64))
+                    || ids.contains(&ValidatorID::Address(validator.pubkey.clone()))
             }
-        }
-    }
-
-    let mut seen: HashSet<u64> = HashSet::new();
-    let mut balances = Vec::new();
-
-    for (i, validator) in validators.iter().enumerate() {
-        let idx = i as u64;
-        let include = filter_ids.is_none()
-            || allowed_indices.contains(&idx)
-            || allowed_addresses.contains(&validator.pubkey);
-
-        if include && seen.insert(idx) {
-            balances.push(ValidatorBalance {
-                index: idx,
-                balance: validator.effective_balance,
-            });
-        }
-    }
-
-    balances
+            None => true,
+        })
+        .map(|(index, (_, balance))| ValidatorBalance {
+            index: index as u64,
+            balance: *balance,
+        })
+        .collect()
 }
 
 #[get("/beacon/states/{state_id}/validator/{validator_id}")]
@@ -373,11 +353,16 @@ pub async fn get_validator_balances_from_state(
         }
     }
 
-    let validator_balances = build_validator_balances(&state.validators, query.id.as_ref());
+    let validator_balances = build_validator_balances(
+        &state
+            .validators
+            .into_iter()
+            .zip(state.balances.into_iter())
+            .collect::<Vec<_>>(),
+        query.id.as_ref(),
+    );
 
-    Ok(HttpResponse::Ok().json(BeaconResponse::new(
-        validator_balances,
-    )))
+    Ok(HttpResponse::Ok().json(BeaconResponse::new(validator_balances)))
 }
 
 #[post("/beacon/states/{state_id}/validator_balances")]
@@ -395,9 +380,14 @@ pub async fn post_validator_balances_from_state(
         }
     }
 
-    let validator_balances = build_validator_balances(&state.validators, body.id.as_ref());
+    let validator_balances = build_validator_balances(
+        &state
+            .validators
+            .into_iter()
+            .zip(state.balances.into_iter())
+            .collect::<Vec<_>>(),
+        body.id.as_ref(),
+    );
 
-    Ok(HttpResponse::Ok().json(BeaconResponse::new(
-        validator_balances,
-    )))
+    Ok(HttpResponse::Ok().json(BeaconResponse::new(validator_balances)))
 }
```

^ here is code that should work then, I think you had an bug because I don't think balance meant effective balance